File: 0290-word-pattern/0290-word-pattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        vector<string> words;
        stringstream ss(s);
        string word;

        while(ss >> word){
            words.push_back(word);
        }
        if(words.size() != pattern.size()) return false;
        unordered_map<string,char>mp1;
        unordered_map<char,string>mp2;

        for(int i =0 ;i< words.size();i++){
            if(mp1.count(words[i]) && mp1[words[i]] != pattern[i]) return false;
            if(mp2.count(pattern[i]) && mp2[pattern[i]]!= words[i]) return false;

            mp1[words[i]] = pattern[i];
            mp2[pattern[i]] = words[i];
        }
        return true;
    }
};

File: 0290-word-pattern/0290-word-pattern.py
class Solution:
    # 2. Word Pattern (LeetCode 290)
    def wordPattern(self, pattern: str, s: str) -> bool:
        # Split the string into individual words
        words = s.split()
        
        # If the number of characters in pattern doesn't match the number of words, return False
        if len(pattern) != len(words):
            return False
            
        # Two dictionaries to maintain a bi-directional mapping between pattern chars and words
        char_to_word = {}
        word_to_char = {}
        
        for c, w in zip(pattern, words):
            # Check for conflict in character-to-word mapping
            if c in char_to_word and char_to_word[c] != w:
                return False
            # Check for conflict in word-to-character mapping
            if w in word_to_char and word_to_char[w] != c:
                return False
                
            # Establish the mapping
            char_to_word[c] = w
            word_to_char[w] = c
            
        # Return True if all mappings are consistent
        return True